semanticdb/metacp/src/main/scala/scala/meta/internal/metacp/asm/JavaTypeSignature.scala

```diff
@@ -0,0 +1,270 @@
+package scala.meta.internal.metacp.asm
+
+import scala.tools.asm.signature.SignatureReader
+
+/** Translation of "Signature" section from the JVM spec to Scala.
+  *
+  * @see https://docs.oracle.com/javase/specs/jvms/se8/html/jvms-4.html#jvms-4.7.9.1
+  */
+sealed trait JavaTypeSignature extends JavaTypeSignature.Printable
+object JavaTypeSignature {
+
+  /** Parse JVM signature using a custom traverser.
+    *
+    * Example:
+    *
+    * {{{
+    *   parse[ClassSignature]("...", new ClassSignatureVisitor)
+    * }}}
+    */
+  final def parse[T](signature: String, visitor: TypedSignatureVisitor[T]): T = {
+    val signatureReader = new SignatureReader(signature)
+    signatureReader.accept(visitor)
+    visitor.result()
+  }
+
+  /** Helper to print parsed JavaTypeSignature back into original signature */
+  trait Printable {
+    def print(sb: StringBuilder): Unit
+    final def pretty: String = {
+      val sb = new StringBuilder
+      this.print(sb)
+      sb.toString
+    }
+  }
```

Comment:
Let's move this trait out of this file. The abstract infrastructure for prettyprinting looks unrelated to the rest of the code.